Document/0x05i-Testing-Code-Quality-and-Build-Settings.md

```diff
@@ -484,88 +484,64 @@ There are various steps to take:
+  - It must be enabled for native code (keywords: C/C++/JNI/NDK) but it might be difficult to fully determine it.
+    - NDK libraries should have it enabled since the compiler does it by default.
+    - Other custom C/C++ libraries might not have it enabled.
+  - If the language is considered memory safe (at least for mitigating buffer overflows), it’s ok if it’s not enabled for that specific binary.
+  - Flutter will not compile using stack canaries because of the way Dart mitigates buffer overflows.
 
 ### Static Analysis
 
-If source code is provided, you can check the build.gradle file to see whether obfuscation settings have been applied. In the example below, you can see that `minifyEnabled` and `proguardFiles` are set. Creating exceptions to protect some classes from obfuscation (with `-keepclassmembers` and `-keep class`) is common. Therefore, auditing the ProGuard configuration file to see what classes are exempted is important. The `getDefaultProguardFile('proguard-android.txt')` method gets the default ProGuard settings from the `<Android SDK>/tools/proguard/` folder.
-
-Further information on how to shrink, obfuscate, and optimize your app can be found in the [Android developer documentation](https://developer.android.com/studio/build/shrink-code "Shrink, obfuscate, and optimize your app").
+Test the app native libraries to determine if they have the PIE an SSP protections enabled.
 
-> When you build you project using Android Studio 3.4 or Android Gradle plugin 3.4.0 or higher, the plugin no longer uses ProGuard to perform compile-time code optimization. Instead, the plugin works with the R8 compiler. R8 works with all of your existing ProGuard rules files, so updating the Android Gradle plugin to use R8 should not require you to change your existing rules.
+You can use rabin2 to get the binary information. We'll use [r2pay-v1.0.apk](https://github.com/OWASP/owasp-mstg/blob/master/Crackmes/Android/Level_04/r2pay-v1.0.apk):
 
-R8 is the new code shrinker from Google and was introduced in Android Studio 3.3 beta. By default, R8 removes attributes that are useful for debugging, including line numbers, source file names, and variable names. R8 is a free Java class file shrinker, optimizer, obfuscator, and pre-verifier and is faster than ProGuard, see also an [Android Developer blog post for further details](https://android-developers.googleblog.com/2018/11/r8-new-code-shrinker-from-google-is.html "R8"). It is shipped with Android's SDK tools. To activate shrinking for the release build, add the following to build.gradle:  
+First we take a look at the main binary for the sake of clarity. You'll see that `canary` and `pic` seem to not enabled (`false`). However, they are actually not applicable, that's why the tool shows them as `false`.
```

Comment:
```suggestion
First, take a look at the main binary. You'll see that `canary` and `pic` are not enabled (`false`). However, they are actually not applicable, which is why the tool shows them as `false`.
```
Here too, if we start the entire testcase with saying "Only test native binaries", this part can become clearer.